File: scripts/upload-sihaenggyuchik.mjs
import { createClient } from '@supabase/supabase-js';
import { readFileSync } from 'fs';
import { resolve } from 'path';
import { homedir } from 'os';

const SUPABASE_URL = 'https://ygohwygdwbckgtotlogm.supabase.co';
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!SUPABASE_KEY) {
  console.error('Set SUPABASE_SERVICE_ROLE_KEY env var');
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

const FILES = [
  {
    path: resolve(homedir(), 'Thairon/legalize-kr/kr/수소경제육성및수소안전관리에관한법률/시행규칙.md'),
    lawName: '수소경제 육성 및 수소 안전관리에 관한 법률 시행규칙',
    lawType: '시행규칙',
  },
  {
    path: resolve(homedir(), 'Thairon/legalize-kr/kr/고압가스안전관리법/시행규칙.md'),
    lawName: '고압가스 안전관리법 시행규칙',
    lawType: '시행규칙',
  },
];

function parseArticles(content, lawName, lawType) {
  // Remove YAML frontmatter
  content = content.replace(/^---\n[\s\S]*?\n---\n/, '');

  const articles = [];
  // Split by article headers: ##### 제N조 or #### 제N조
  const pattern = /(#{4,5}\s+제\d+조(?:의\d+)?\s*(?:\([^)]*\))?)/g;
  const headers = [];
  let match;

  while ((match = pattern.exec(content)) !== null) {
    headers.push({ index: match.index, header: match[1].trim() });
  }

  for (let i = 0; i < headers.length; i++) {
    const header = headers[i].header;
    const start = headers[i].index + headers[i].header.length;
    const end = i + 1 < headers.length ? headers[i + 1].index : content.length;
    const body = content.slice(start, end).trim();

    // Extract article number
    const numMatch = header.match(/제(\d+)조(?:의(\d+))?/);
    let articleNo = numMatch ? `제${numMatch[1]}조` : header;
    if (numMatch && numMatch[2]) articleNo += `의${numMatch[2]}`;

    // Extract title
    const titleMatch = header.match(/\(([^)]+)\)/);
    const title = titleMatch ? titleMatch[1] : '';

    const fullContent = `${header}\n\n${body}`;
    const cleanName = lawName.replace(/\s+/g, '');
    const id = `${cleanName}_${articleNo}`;

    articles.push({ id, law_name: lawName, law_id: null, article_no: articleNo, title, content: fullContent, law_type: lawType });
  }

  return articles;
}

async function main() {
  let totalUploaded = 0;

  for (const file of FILES) {
    let content;
    try {
      content = readFileSync(file.path, 'utf-8');
    } catch (e) {
      console.log(`File not found: ${file.path}`);
      continue;
    }

    const articles = parseArticles(content, file.lawName, file.lawType);
    console.log(`\n${file.lawName}: ${articles.length} articles parsed`);

    let uploaded = 0;
    for (const article of articles) {
      try {
        const { error } = await supabase.from('law_articles').upsert(article, { onConflict: 'id' });
        if (error) {
          console.log(`  Error: ${article.id} - ${error.message}`);
        } else {
          uploaded++;
        }
      } catch (e) {
        console.log(`  Error: ${article.id} - ${e.message}`);
      }
    }

    console.log(`  Uploaded: ${uploaded}/${articles.length}`);
    totalUploaded += uploaded;
  }

  console.log(`\nTotal uploaded: ${totalUploaded}`);
}

main();

File: scripts/upload_sihaenggyuchik.py
"""Upload 시행규칙 from legalize-kr to Supabase law_articles table"""

import os
import re
import json
from supabase import create_client

SUPABASE_URL = os.environ.get("NEXT_PUBLIC_SUPABASE_URL") or os.environ.get("SUPABASE_URL")
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY") or os.environ.get("SUPABASE_KEY")

FILES = [
    {
        "path": os.path.expanduser("~/Thairon/legalize-kr/kr/수소경제육성및수소안전관리에관한법률/시행규칙.md"),
        "law_name": "수소경제 육성 및 수소 안전관리에 관한 법률 시행규칙",
        "law_type": "시행규칙",
    },
    {
        "path": os.path.expanduser("~/Thairon/legalize-kr/kr/고압가스안전관리법/시행규칙.md"),
        "law_name": "고압가스 안전관리법 시행규칙",
        "law_type": "시행규칙",
    },
]


def parse_articles(content: str, law_name: str, law_type: str):
    """Parse markdown into individual articles"""
    # Remove YAML frontmatter
    content = re.sub(r'^---\n.*?\n---\n', '', content, flags=re.DOTALL)

    # Split by article headers: ##### 제N조 (...)
    pattern = r'(#{4,5}\s+제\d+조(?:의\d+)?\s*(?:\([^)]*\))?)'
    parts = re.split(pattern, content)

    articles = []
    i = 1  # skip content before first article
    while i < len(parts):
        header = parts[i].strip()
        body = parts[i + 1] if i + 1 < len(parts) else ''

        # Extract article number
        num_match = re.search(r'제(\d+)조(?:의(\d+))?', header)
        if num_match:
            article_no = f"제{num_match.group(1)}조"
            if num_match.group(2):
                article_no += f"의{num_match.group(2)}"
        else:
            article_no = header

        # Extract title from parentheses
        title_match = re.search(r'\(([^)]+)\)', header)
        title = title_match.group(1) if title_match else ''

        # Clean up body
        full_content = f"{header}\n\n{body.strip()}"

        # Generate ID
        clean_name = re.sub(r'\s+', '', law_name)
        article_id = f"{clean_name}_{article_no}"

        articles.append({
            "id": article_id,
            "law_name": law_name,
            "law_id": None,
            "article_no": article_no,
            "title": title,
            "content": full_content,
            "law_type": law_type,
        })

        i += 2

    return articles


def main():
    if not SUPABASE_URL or not SUPABASE_KEY:
        print("Error: SUPABASE_URL and SUPABASE_KEY environment variables required")
        print("Set NEXT_PUBLIC_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY")
        return

    supabase = create_client(SUPABASE_URL, SUPABASE_KEY)

    total_uploaded = 0

    for file_info in FILES:
        path = file_info["path"]
        if not os.path.exists(path):
            print(f"File not found: {path}")
            continue

        with open(path, 'r', encoding='utf-8') as f:
            content = f.read()

        articles = parse_articles(content, file_info["law_name"], file_info["law_type"])
        print(f"\n{file_info['law_name']}: {len(articles)} articles parsed")

        # Upload one by one to handle duplicates
        uploaded = 0
        for article in articles:
            try:
                supabase.table("law_articles").upsert(article, on_conflict="id").execute()
                uploaded += 1
            except Exception as e:
                print(f"  Error uploading {article['id']}: {e}")

        print(f"  Uploaded: {uploaded}/{len(articles)}")
        total_uploaded += uploaded

    print(f"\nTotal uploaded: {total_uploaded}")


if __name__ == "__main__":
    main()
